### PR TITLE
Issue 12829 - Wrong error line number for closure allocation in @nogc function

### PR DESCRIPTION
**Pull request by 9rnsr**
https://issues.dlang.org/show_bug.cgi?id=12829

### CODE REVIEW

**Comment by yebblies** [comment]
This PR looks OK, but I'm finding the control flow quite hard to follow.

Would you consider:
- Removing a level of nesting by returning false immediately if `needsClosure` returns `false`.
- Moving the `k` loop inside the `s` loop and using labelled `continue` instead of `goto`
- Move the insert into `a` out of the `k` loop

My hope is these changes would make it easier to see what's going on.

The `i` and `j` loops could also make use of `foreach`, although that's less important.

File: src/func.d [hunk 1]
```diff
@@ -2169,13 +2169,12 @@ public:
...
         }
-        if (needsClosure())
+
+        if (checkClosure())
         {
-            if (setGC())
-                error("@nogc function allocates a closure with the GC");
-            else
-                printGCUsage(loc, "using closure causes GC allocation");
+            //errors = true;
```
Comment (yebblies) [inline src/func.d]: What does this comment mean?

**Comment by 9rnsr** [comment]
@yebblies Imrpove the control flow in `FuncDeclaration.checkClosure()` as you suggested.

File: src/func.d (hunk 1 above)
Comment (9rnsr) [inline src/func.d]: Currently all error occurrences in `FuncDeclaration.semantic3` is reflected into `FuncDeclaration.errors` variable by using `global.errors` at the end of `semantic3`. But, Walter is continuously saying that the it's a bad design and should be removed in the future. I think too.
Therefore, although today the line `errors = true;` is just a dead code, but it would be enabled when we'll finally remove `global.errors`.

Comment (yebblies) [inline src/func.d]: Could you put a short version of that explanation in the comment instead?
eg `// We should be setting errors here instead of relying on the global error count`
The meaning of dead code as comments unfortunately tends to get lost over time.

**Comment by yebblies** [comment]
> @yebblies Imrpove the control flow in FuncDeclaration.checkClosure() as you suggested.

That is much easier to follow, thanks.  It's surprising how much nicer it is with foreach.

File: src/func.d (hunk 1 above)
Comment (9rnsr) [inline src/func.d]: Nice one-liner comment. I added it here.

**Comment by yebblies** [comment]
Auto-merge toggled on

**Comment by 9rnsr** [comment]
Thanks!

File: src/func.d [hunk 2]
```diff
@@ -3497,6 +3496,62 @@ public:
     }
 
     /***********************************************
+     * Check that the function contains any closure.
+     * If it's @nogc, report suitable errors.
+     * This is mostly consisten with FuncDeclaration::needsClosure().
```
Comment (WalterBright) [inline src/func.d]: /consisten/consistent/

Comment (WalterBright) [inline src/func.d]: Being in D now, need to use Ddoc comments for functions. In particular, the return value is not documented with a `Returns:` block.

**Comment by WalterBright** [comment]
Auto-merge toggled off

**Comment by WalterBright** [comment]
Please turn auto merge back on after the comment is fixed. Thanks!

**Comment by 9rnsr** [comment]
I fixed a miss-spelling in the comment, then added `Returns:` block for Ddoc generation.

**Comment by 9rnsr** [comment]
Auto-merge toggled on

**Comment by 9rnsr** [comment]
Turn on auto-merge again.

**Comment by WalterBright** [comment]
thanks
